### PR TITLE
Add features flag support to parse options

### PR DESCRIPTION
**Pull request by khyperia**
CSharpParseOptions was hardcoded to reject any /features flags passed in, and any unit tests that used feature flags. This PR changes that, which is required for any future C#7 experimental language features, as well as use sites of feature flags.

There is a stub in MessageID.cs where feature flag strings can be defined.

@gafter @jaredpar @VSadov @AlekseyTs @agocke Please review.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/CSharpParseOptions.cs [hunk 1]
```diff
@@ -18,6 +18,8 @@ public sealed class CSharpParseOptions : ParseOptions, IEquatable<CSharpParseOpt
         /// </summary>
         public static readonly CSharpParseOptions Default = new CSharpParseOptions();
 
+        private IReadOnlyDictionary<string, string> _features;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CSharpParseOptions.cs]: Can we make this readonly?  

If this is only ever assigned `ImmutableDictionary` can we make it that type?

File: src/Compilers/CSharp/Portable/CSharpParseOptions.cs [hunk 2]
```diff
@@ -178,26 +180,30 @@ public new CSharpParseOptions WithFeatures(IEnumerable<KeyValuePair<string, stri
...
 
         public override IReadOnlyDictionary<string, string> Features
         {
             get
             {
-                // there are currently no parse options for experimental features
-                return new Dictionary<string, string>();
+                if (_features == null)
+                {
+                    _features = new Dictionary<string, string>();
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CSharpParseOptions.cs]: Use the following:

```
_features = ImmutableDictionary<string, string>.Empty
```

It avoids an allocation. 

Note: we could simplify this to just make `_features` default to this value.

File: src/Compilers/CSharp/Portable/Errors/MessageID.cs [hunk 3]
```diff
@@ -145,6 +145,15 @@ public static LocalizableErrorArgument Localize(this MessageID id)
             return new LocalizableErrorArgument(id);
         }
 
+        internal static string RequiredFeature(this MessageID feature)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Errors/MessageID.cs]: Nit: would named this to indicate an action is occuring.  For example ConvertToRequiredFeatureId.

File: src/Compilers/CSharp/Portable/CSharpParseOptions.cs (hunk 1 above)
Comment (khyperia) [inline src/Compilers/CSharp/Portable/CSharpParseOptions.cs]: It can't be readonly due to WithLanguageVersion assigning it after construction. WithLanguageVersion can't pass it into the constructor, either, because it's a public API and I assume shouldn't be changed. (CompilationOptions does the same thing, with a mutable field hidden under a With* method)

The only place it should be used is in the Features property, which is of type IReadOnlyDictionary, so I figured it should be of that type.

Comment (agocke) [inline src/Compilers/CSharp/Portable/CSharpParseOptions.cs]: I think we should make it `ImmutableDictionary` on the philosophy "be specific in what you provide and lenient in what you accept."

Of course, with the caveat that in perf optimization cases my rule changes to "whatever will make it faster."

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs [hunk 4]
```diff
@@ -1914,7 +1914,17 @@ private AssemblySymbol GetForwardedToAssembly(string fullName, int arity, out bo
 
         internal static void CheckFeatureAvailability(Location location, MessageID feature, DiagnosticBag diagnostics)
         {
-            LanguageVersion availableVersion = ((CSharpParseOptions)location.SourceTree.Options).LanguageVersion;
+            var options = (CSharpParseOptions)location.SourceTree.Options;
+            string featureFlag = feature.RequiredFeature();
+            if (featureFlag != null)
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: Is this the only usage of `featureFlag`? If so, could you just inline the expression?

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs [hunk 5]
```diff
@@ -1914,7 +1914,17 @@ private AssemblySymbol GetForwardedToAssembly(string fullName, int arity, out bo
 
         internal static void CheckFeatureAvailability(Location location, MessageID feature, DiagnosticBag diagnostics)
         {
-            LanguageVersion availableVersion = ((CSharpParseOptions)location.SourceTree.Options).LanguageVersion;
+            var options = (CSharpParseOptions)location.SourceTree.Options;
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: Is the feature flag still a parse-only option? What are we doing for features which have new semantics but no new syntax (e.g., get-only autoprop initialization in C# 6 had this property).

File: src/Compilers/CSharp/Portable/CSharpParseOptions.cs (hunk 2 above)
Comment (agocke) [inline src/Compilers/CSharp/Portable/CSharpParseOptions.cs]: If it's not in the initializer, do we need double-checked locking here or interlocked compareexchange?

Comment (khyperia) [inline src/Compilers/CSharp/Portable/CSharpParseOptions.cs]: @agocke I changed it to what @jaredpar mentioned, putting ImmutableDictionary.Empty in the constructor. I didn't know about Empty, and did it this lazy way to avoid an allocation if it was just going to be overwritten in WithFeatures anyway. Avoids the interlocked problem that way.

Comment (agocke) [inline src/Compilers/CSharp/Portable/CSharpParseOptions.cs]: @khyperia Cool -- feel free to just update the PR with your latest changeset. GitHub will mark my comment as obsolete if the diff changes.

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs (hunk 5 above)
Comment (khyperia) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: Not sure. I do know that this method is called in a few places in a place for non-syntax-features, e.g. https://github.com/khyperia/roslyn/blob/dff492c59ac5a2da70d89b32b5f9d12799bf8fd0/src/Compilers/CSharp/Portable/Binder/SwitchBinder.cs#L344

Also, CSharpCompilationOptions.Features contains the same items as CSharpParseOptions.Features, so if a CompilationOptions is in scope somewhere, that might be better to grab than ParseOptions?

**Comment by khyperia** [comment]
@gafter @jaredpar @VSadov @AlekseyTs @agocke Wasn't sure if updating a commit notified people, but I did so a bit back - could I have a re-review?

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs (hunk 5 above)
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: I think @gafter designed this section so we should probably have him weigh in here on its intended usage strategy.

From what I can see the feature gating on syntax and compilation are separate, but interacting. In parsing I would expect only the parse flags on the parse options to be checked, but here, in the binder, I would expect both the parse options AND the compilation options to be checked in case the feature has either new syntax or new semantics.

I could also see the argument that only the compilation options need to be checked because presumably the parser would have gone through and added any parse-level feature diagnostics already.

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs [hunk 6]
```diff
@@ -893,6 +893,16 @@ private void ScanSyntaxToken(ref TokenInfo info)
 
         private void CheckFeatureAvailability(MessageID feature)
         {
+            string featureFlag = feature.RequiredFeature();
+            if (featureFlag != null)
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Parser/Lexer.cs]: Inline `featureFlag` here as well?

File: src/Compilers/CSharp/Portable/CommandLine/CommandLineParser.cs [hunk 7]
```diff
@@ -1039,7 +1039,7 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
                 preprocessorSymbols: defines.ToImmutableAndFree(),
                 documentationMode: parseDocumentationComments ? DocumentationMode.Diagnose : DocumentationMode.None,
                 kind: SourceCodeKind.Regular
-            );
+            ).WithFeatures(features.ToImmutableDictionary(feature => feature, feature => "true"));
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CommandLine/CommandLineParser.cs]: We should add an overload, even if it's only internal, which takes features as a parameter.  As written we're unnecessarily allocating `CSharpParseOption` instances

Comment (khyperia) [inline src/Compilers/CSharp/Portable/CommandLine/CommandLineParser.cs]: Should that go into a separate PR? We're doing the same thing for CSharpCompilationOptions, and I would guess that both should be updated (or this one could grow). VB should get the same change too (but VB also needs the equivalent of this PR applied to its parse options).

https://github.com/khyperia/roslyn/blob/FeaturesFlag/src/Compilers/CSharp/Portable/CommandLine/CommandLineParser.cs#L1064

Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CommandLine/CommandLineParser.cs]: @khyperia given that this PR is already changing `CSharpParseOptions` I would just go ahead and do it now.

**Comment by AlekseyTs** [comment]
Consider adding more unit-tests.

**Comment by AlekseyTs** [comment]
Rather than deleting/modifying old code, consider if it is worth conditionally disabling it by using ```#if``` in order to make it simpler to remove this support during preparation to the next release.

**Comment by khyperia** [comment]
@jaredpar Updated to add features parameter to CSharpParseOptions/CompilationOptions as an internal constructor. All other use sites of WithFeatures other than in the CommandLineParser (which I updated to use the new constructor) are in tests.

@AlekseyTs I'm not sure what to add unit tests for, as the features flag is unused in the compiler in this branch (and so is difficult to measure the effects of) - it's just stored in the options classes now, which I'm testing.

Not sure what you mean by the `#if` comment - what is being conditionally disabled? Rejecting any calls to WithFeatures, or checking feature flags alongside language version checks, both, or what?

File: src/Compilers/CSharp/Portable/Errors/MessageID.cs (hunk 3 above)
Comment (khyperia) [inline src/Compilers/CSharp/Portable/Errors/MessageID.cs]: `RequiredVersion` is just below this with a similar name (although that could be changed too). For reference, VB does `GetLanguageVersion` (and I'm about to submit a PR that adds `GetFeatureFlag`). I feel like `RequiredFeature` reads better, although I think your example name sounds like a conversion from a feature flag string to a MessageID, which it's the opposite.

**Comment by jaredpar** [comment]
Looks like the build is failing.  Is it for the same reason as the VB change?  If so I would just hand port the fix to the failing test into this PR

**Comment by khyperia** [comment]
@jaredpar Yes, looking at the logs, it looks like the test that failed is ScriptCompilationInternalOnly, which is the same one as the VB one. The "fix" (deleting the test) should be automatically transferred from master to future tonight. We could do a manual PR from master to future to speed up the process.

File: src/Compilers/CSharp/Portable/CSharpCompilationOptions.cs [hunk 8]
```diff
@@ -65,6 +65,44 @@ public sealed class CSharpCompilationOptions : CompilationOptions, IEquatable<CS
         {
         }
 
+        internal CSharpCompilationOptions(
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/CSharpCompilationOptions.cs]: We already have constructors with the `features` flag. Why add a new one?

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs (hunk 5 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: When creating a `CSharpCompilation`, the caller is responsible for ensuring that all trees were constructed using the same language settings (`ArgumentException` is thrown by the constructor if they disagree), and that is the language version assigned to the compilation in the constructor.

I suggest we should remove `features` from `CSharpCompilationOptions` and do the same thing for the features setting.

Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: @gafter it's a `public` API.  I sent mail about this today.

File: src/Compilers/CSharp/Portable/CSharpCompilationOptions.cs (hunk 8 above)
Comment (khyperia) [inline src/Compilers/CSharp/Portable/CSharpCompilationOptions.cs]: Usages of the constructor that use the features parameter (i.e. CommandLineParser) don't provide all the parameters of the constructor, letting some be their default parameters. The one with the features parameter doesn't have optional parameters, so I'd have to extract the defaults into the CommandLineParser to be able to use that constructor.

File: src/Compilers/CSharp/Portable/CommandLine/CommandLineParser.cs [hunk 9]
```diff
@@ -1038,7 +1038,8 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
                 languageVersion: languageVersion,
                 preprocessorSymbols: defines.ToImmutableAndFree(),
                 documentationMode: parseDocumentationComments ? DocumentationMode.Diagnose : DocumentationMode.None,
-                kind: SourceCodeKind.Regular
+                kind: SourceCodeKind.Regular,
+                features: features.ToImmutableDictionary(feature => feature, feature => "true")
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/CommandLine/CommandLineParser.cs]: Actually, you should look for an `=` in the string, so that you can set `/feature:foo=bar`. You should only use `true` when there is no `=`.

Comment (khyperia) [inline src/Compilers/CSharp/Portable/CommandLine/CommandLineParser.cs]: The command line parser currently doesn't do much /feature parameter parsing. Expected behavior is specified in a test, this assert in particular is interesting.

https://github.com/dotnet/roslyn/blob/80d2e0bf26ef5eddea43b1e9b506e8b604bf5db5/src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs#L6856

I think @jaredpar might have more to say on this (with respect to having an ImmutableArray in CompilationOptions and Dictionary in ParseOptions)

**Comment by jaredpar** [comment]
:+1: 

As discussed we will deal with the `ImmutableArray<string>` changes and general usage within `CompilationOptions` with a separate PR.

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs (hunk 5 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: We'll take that to shiproom today.

File: src/Compilers/CSharp/Portable/CommandLine/CommandLineParser.cs (hunk 9 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/CommandLine/CommandLineParser.cs]: My mistake; you should be looking for a `:` (colon) separating the key and value.

**Comment by gafter** [comment]
LGTM, with follow-up expected in `CompilationOptions` separately (some of which will probably flow down from the stabilization branch).
